// @enact/sandstone/MediaPlayer/MediaControls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MediaControlsDecorator = exports.MediaControls = exports.MediaControlsBase = exports["default"] = void 0;

var _ApiDecorator = _interopRequireDefault(require("@enact/core/internal/ApiDecorator"));

var _Cancelable = _interopRequireDefault(require("@enact/ui/Cancelable"));

var _kind = _interopRequireDefault(require("@enact/core/kind"));

var _hoc = _interopRequireDefault(require("@enact/core/hoc"));

var _keymap = require("@enact/core/keymap");

var _dispatcher = require("@enact/core/dispatcher");

var _Pause = _interopRequireDefault(require("@enact/spotlight/Pause"));

var _Slottable = _interopRequireDefault(require("@enact/ui/Slottable"));

var _spotlight = _interopRequireDefault(require("@enact/spotlight"));

var _SpotlightContainerDecorator = require("@enact/spotlight/SpotlightContainerDecorator");

var _handle = require("@enact/core/handle");

var _util = require("@enact/core/util");

var _onlyUpdateForKeys = _interopRequireDefault(require("recompose/onlyUpdateForKeys"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = require("react");

var _reactDom = _interopRequireDefault(require("react-dom"));

var _$L = _interopRequireDefault(require("../internal/$L"));

var _util2 = require("../internal/util");

var _ActionGuide = _interopRequireDefault(require("../ActionGuide"));

var _Button = _interopRequireDefault(require("../Button"));

var _util3 = require("./util");

var _MediaControlsModule = _interopRequireDefault(require("./MediaControls.module.css"));

var _jsxRuntime = require("react/jsx-runtime");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var OuterContainer = (0, _SpotlightContainerDecorator.SpotlightContainerDecorator)({
  defaultElement: [".".concat(_SpotlightContainerDecorator.spotlightDefaultClass)]
}, 'div');
var Container = (0, _SpotlightContainerDecorator.SpotlightContainerDecorator)({
  enterTo: 'default-element'
}, 'div');
var MediaButton = (0, _onlyUpdateForKeys["default"])(['children', 'className', 'disabled', 'icon', 'onClick', 'spotlightDisabled'])(_Button["default"]);
var forwardToggleMore = (0, _handle.forward)('onToggleMore');
var animationDuration = 300;
/**
 * A set of components for controlling media playback and rendering additional components.
 *
 * @class MediaControlsBase
 * @memberof sandstone/MediaPlayer
 * @ui
 * @private
 */

var MediaControlsBase = (0, _kind["default"])({
  name: 'MediaControls',
  // intentionally assigning these props to MediaControls instead of Base (which is private)
  propTypes:
  /** @lends sandstone/MediaPlayer.MediaControls.prototype */
  {
    /**
     * DOM id for the component.
     *
     * This child component `ActionGuide`'s id is generated from the id.
     *
     * @type {String}
     * @required
     * @public
     */
    id: _propTypes["default"].string.isRequired,

    /**
     * The `aria-label` for the action guide.
     *
     * @type {String}
     * @public
     */
    actionGuideAriaLabel: _propTypes["default"].string,

    /**
     * The label for the action guide.
     *
     * @type {String}
     * @public
     */
    actionGuideLabel: _propTypes["default"].string,

    /**
     * These components are placed below the action guide. Typically these will be media playlist controls.
     *
     * @type {Node}
     * @public
     */
    bottomComponents: _propTypes["default"].node,

    /**
     * Jump backward [icon]{@link sandstone/Icon.Icon} name. Accepts any
     * [icon]{@link sandstone/Icon.Icon} component type.
     *
     * @type {String}
     * @default 'jumpbackward'
     * @public
     */
    jumpBackwardIcon: _propTypes["default"].string,

    /**
     * Disables state on the media "jump" buttons; the outer pair.
     *
     * @type {Boolean}
     * @public
     */
    jumpButtonsDisabled: _propTypes["default"].bool,

    /**
     * Jump forward [icon]{@link sandstone/Icon.Icon} name. Accepts any
     * [icon]{@link sandstone/Icon.Icon} component type.
     *
     * @type {String}
     * @default 'jumpforward'
     * @public
     */
    jumpForwardIcon: _propTypes["default"].string,

    /**
     * Disables the media buttons.
     *
     * @type {Boolean}
     * @public
     */
    mediaDisabled: _propTypes["default"].bool,

    /**
     * When `true`, more components are rendered. This does not indicate the visibility of more components.
     *
     * @type {Boolean}
     * @public
     */
    moreComponentsRendered: _propTypes["default"].bool,

    /**
     * The spotlight ID for the moreComponent container.
     *
     * @type {String}
     * @public
     * @default 'moreComponents'
     */
    moreComponentsSpotlightId: _propTypes["default"].string,

    /**
     * Removes the "jump" buttons. The buttons that skip forward or backward in the video.
     *
     * @type {Boolean}
     * @public
     */
    noJumpButtons: _propTypes["default"].bool,

    /**
     * Called when cancel/back key events are fired.
     *
     * @type {Function}
     * @public
     */
    onClose: _propTypes["default"].func,

    /**
     * Called when the user flicks on the action guide.
     *
     * @type {Function}
     * @private
     */
    onFlickFromActionGuide: _propTypes["default"].func,

    /**
     * Called when the user clicks the JumpBackward button
     *
     * @type {Function}
     * @public
     */
    onJumpBackwardButtonClick: _propTypes["default"].func,

    /**
     * Called when the user clicks the JumpForward button.
     *
     * @type {Function}
     * @public
     */
    onJumpForwardButtonClick: _propTypes["default"].func,

    /**
     * Called when the user presses a media control button.
     *
     * @type {Function}
     * @public
     */
    onKeyDownFromMediaButtons: _propTypes["default"].func,

    /**
     * Called when the user clicks the Play button.
     *
     * @type {Function}
     * @public
     */
    onPlayButtonClick: _propTypes["default"].func,

    /**
     * `true` when the video is paused.
     *
     * @type {Boolean}
     * @public
     */
    paused: _propTypes["default"].bool,

    /**
     * A string which is sent to the `pause` icon of the player controls. This can be
     * anything that is accepted by [Icon]{@link sandstone/Icon.Icon}. This will be temporarily replaced by
     * the [playIcon]{@link sandstone/MediaPlayer.MediaControls.playIcon} when the
     * [paused]{@link sandstone/MediaPlayer.MediaControls.paused} boolean is `false`.
     *
     * @type {String}
     * @default 'pause'
     * @public
     */
    pauseIcon: _propTypes["default"].string,

    /**
     * A string which is sent to the `play` icon of the player controls. This can be
     * anything that is accepted by {@link sandstone/Icon.Icon}. This will be temporarily replaced by
     * the [pauseIcon]{@link sandstone/MediaPlayer.MediaControls.pauseIcon} when the
     * [paused]{@link sandstone/MediaPlayer.MediaControls.paused} boolean is `true`.
     *
     * @type {String}
     * @default 'play'
     * @public
     */
    playIcon: _propTypes["default"].string,

    /**
     * Disables the media "play"/"pause" button.
     *
     * @type {Boolean}
     * @public
     */
    playPauseButtonDisabled: _propTypes["default"].bool,

    /**
     * When `true`, more components are visible.
     *
     * @type {Boolean}
     * @private
     */
    showMoreComponents: _propTypes["default"].bool,

    /**
     * `true` controls are disabled from Spotlight.
     *
     * @type {Boolean}
     * @public
     */
    spotlightDisabled: _propTypes["default"].bool,

    /**
     * The spotlight ID for the media controls container.
     *
     * @type {String}
     * @public
     * @default 'mediaControls'
     */
    spotlightId: _propTypes["default"].string,

    /**
     * The visibility of the component. When `false`, the component will be hidden.
     *
     * @type {Boolean}
     * @default true
     * @public
     */
    visible: _propTypes["default"].bool
  },
  defaultProps: {
    jumpBackwardIcon: 'jumpbackward',
    jumpForwardIcon: 'jumpforward',
    moreComponentsSpotlightId: 'moreComponents',
    spotlightId: 'mediaControls',
    pauseIcon: 'pause',
    playIcon: 'play',
    visible: true
  },
  styles: {
    css: _MediaControlsModule["default"],
    className: 'controlsFrame'
  },
  computed: {
    actionGuideClassName: function actionGuideClassName(_ref) {
      var styler = _ref.styler,
          showMoreComponents = _ref.showMoreComponents;
      return styler.join({
        hidden: showMoreComponents
      });
    },
    actionGuideShowing: function actionGuideShowing(_ref2) {
      var bottomComponents = _ref2.bottomComponents,
          children = _ref2.children;
      return (0, _util3.countReactChildren)(children) || bottomComponents;
    },
    className: function className(_ref3) {
      var visible = _ref3.visible,
          styler = _ref3.styler;
      return styler.append({
        hidden: !visible
      });
    },
    moreButtonsClassName: function moreButtonsClassName(_ref4) {
      var styler = _ref4.styler;
      return styler.join('mediaControls', 'moreButtonsComponents');
    },
    moreComponentsRendered: function moreComponentsRendered(_ref5) {
      var showMoreComponents = _ref5.showMoreComponents,
          _moreComponentsRendered = _ref5.moreComponentsRendered;
      return showMoreComponents || _moreComponentsRendered;
    }
  },
  render: function render(_ref6) {
    var actionGuideAriaLabel = _ref6.actionGuideAriaLabel,
        actionGuideLabel = _ref6.actionGuideLabel,
        actionGuideShowing = _ref6.actionGuideShowing,
        children = _ref6.children,
        id = _ref6.id,
        jumpBackwardIcon = _ref6.jumpBackwardIcon,
        jumpButtonsDisabled = _ref6.jumpButtonsDisabled,
        jumpForwardIcon = _ref6.jumpForwardIcon,
        bottomComponents = _ref6.bottomComponents,
        mediaDisabled = _ref6.mediaDisabled,
        moreComponentsSpotlightId = _ref6.moreComponentsSpotlightId,
        noJumpButtons = _ref6.noJumpButtons,
        onFlickFromActionGuide = _ref6.onFlickFromActionGuide,
        onJumpBackwardButtonClick = _ref6.onJumpBackwardButtonClick,
        onJumpForwardButtonClick = _ref6.onJumpForwardButtonClick,
        onKeyDownFromMediaButtons = _ref6.onKeyDownFromMediaButtons,
        onPlayButtonClick = _ref6.onPlayButtonClick,
        paused = _ref6.paused,
        pauseIcon = _ref6.pauseIcon,
        playIcon = _ref6.playIcon,
        playPauseButtonDisabled = _ref6.playPauseButtonDisabled,
        showMoreComponents = _ref6.showMoreComponents,
        moreComponentsRendered = _ref6.moreComponentsRendered,
        moreButtonsClassName = _ref6.moreButtonsClassName,
        actionGuideClassName = _ref6.actionGuideClassName,
        spotlightDisabled = _ref6.spotlightDisabled,
        spotlightId = _ref6.spotlightId,
        rest = _objectWithoutProperties(_ref6, ["actionGuideAriaLabel", "actionGuideLabel", "actionGuideShowing", "children", "id", "jumpBackwardIcon", "jumpButtonsDisabled", "jumpForwardIcon", "bottomComponents", "mediaDisabled", "moreComponentsSpotlightId", "noJumpButtons", "onFlickFromActionGuide", "onJumpBackwardButtonClick", "onJumpForwardButtonClick", "onKeyDownFromMediaButtons", "onPlayButtonClick", "paused", "pauseIcon", "playIcon", "playPauseButtonDisabled", "showMoreComponents", "moreComponentsRendered", "moreButtonsClassName", "actionGuideClassName", "spotlightDisabled", "spotlightId"]);

    delete rest.onClose;
    delete rest.visible;
    return /*#__PURE__*/(0, _jsxRuntime.jsxs)(OuterContainer, _objectSpread(_objectSpread({}, rest), {}, {
      id: id,
      spotlightId: spotlightId,
      children: [/*#__PURE__*/(0, _jsxRuntime.jsxs)(Container, {
        className: _MediaControlsModule["default"].mediaControls,
        spotlightDisabled: spotlightDisabled,
        onKeyDown: onKeyDownFromMediaButtons,
        children: [noJumpButtons ? null : /*#__PURE__*/(0, _jsxRuntime.jsx)(MediaButton, {
          "aria-label": (0, _$L["default"])('Previous'),
          backgroundOpacity: "transparent",
          css: _MediaControlsModule["default"],
          disabled: mediaDisabled || jumpButtonsDisabled,
          icon: jumpBackwardIcon,
          onClick: onJumpBackwardButtonClick,
          size: "large",
          spotlightDisabled: spotlightDisabled
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)(MediaButton, {
          "aria-label": paused ? (0, _$L["default"])('Play') : (0, _$L["default"])('Pause'),
          className: _SpotlightContainerDecorator.spotlightDefaultClass,
          backgroundOpacity: "transparent",
          css: _MediaControlsModule["default"],
          disabled: mediaDisabled || playPauseButtonDisabled,
          icon: paused ? playIcon : pauseIcon,
          onClick: onPlayButtonClick,
          size: "large",
          spotlightDisabled: spotlightDisabled
        }), noJumpButtons ? null : /*#__PURE__*/(0, _jsxRuntime.jsx)(MediaButton, {
          "aria-label": (0, _$L["default"])('Next'),
          backgroundOpacity: "transparent",
          css: _MediaControlsModule["default"],
          disabled: mediaDisabled || jumpButtonsDisabled,
          icon: jumpForwardIcon,
          onClick: onJumpForwardButtonClick,
          size: "large",
          spotlightDisabled: spotlightDisabled
        })]
      }), actionGuideShowing ? /*#__PURE__*/(0, _jsxRuntime.jsx)(_ActionGuide["default"], {
        id: "".concat(id, "_actionGuide"),
        "aria-label": actionGuideAriaLabel != null ? actionGuideAriaLabel : null,
        css: _MediaControlsModule["default"],
        className: actionGuideClassName,
        icon: "arrowsmalldown",
        onFlick: onFlickFromActionGuide,
        children: actionGuideLabel
      }) : null, moreComponentsRendered ? /*#__PURE__*/(0, _jsxRuntime.jsxs)(Container, {
        spotlightId: moreComponentsSpotlightId,
        className: _MediaControlsModule["default"].moreComponents,
        spotlightDisabled: !showMoreComponents || spotlightDisabled,
        children: [/*#__PURE__*/(0, _jsxRuntime.jsx)(Container, {
          className: moreButtonsClassName,
          children: children
        }), /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {
          children: bottomComponents
        })]
      }) : null]
    }));
  }
});
/**
 * Media control behaviors to apply to [MediaControlsBase]{@link sandstone/MediaPlayer.MediaControlsBase}.
 * Provides built-in support for showing more components and key handling for basic playback
 * controls.
 *
 * @class MediaControlsDecorator
 * @memberof sandstone/MediaPlayer
 * @mixes ui/Slottable.Slottable
 * @hoc
 * @private
 */

exports.MediaControlsBase = MediaControlsBase;
var MediaControlsDecorator = (0, _hoc["default"])(function (config, Wrapped) {
  var MediaControlsDecoratorHOC = /*#__PURE__*/function (_Component) {
    _inherits(MediaControlsDecoratorHOC, _Component);

    var _super = _createSuper(MediaControlsDecoratorHOC);

    function MediaControlsDecoratorHOC(props) {
      var _this;

      _classCallCheck(this, MediaControlsDecoratorHOC);

      _this = _super.call(this, props);
      _this.moreComponentsRenderingJob = new _util.Job(function () {
        _this.setState({
          moreComponentsRendered: true
        });
      });

      _this.calculateMoreComponentsHeight = function () {
        if (!_this.mediaControlsNode) {
          _this.bottomComponentsHeight = 0;
          return;
        }

        var bottomElement = _this.mediaControlsNode.querySelector(".".concat(_MediaControlsModule["default"].moreComponents));

        _this.bottomComponentsHeight = bottomElement ? bottomElement.scrollHeight : 0;
      };

      _this.canShowMoreComponents = function () {
        return !_this.props.moreActionDisabled && !_this.state.showMoreComponents;
      };

      _this.handleKeyDownFromMediaButtons = function (ev) {
        if ((0, _keymap.is)('down', ev.keyCode) && _this.canShowMoreComponents()) {
          _this.showMoreComponents();

          ev.stopPropagation();
        }
      };

      _this.handleFlickFromActionGuide = function (_ref7) {
        var direction = _ref7.direction,
            velocityY = _ref7.velocityY;

        if (direction === 'vertical' && velocityY < 0 && _this.canShowMoreComponents()) {
          _this.showMoreComponents();
        }
      };

      _this.handleKeyDown = function (ev) {
        var _this$props = _this.props,
            mediaDisabled = _this$props.mediaDisabled,
            no5WayJump = _this$props.no5WayJump,
            visible = _this$props.visible;

        var current = _spotlight["default"].getCurrent();

        if (!no5WayJump && !visible && !mediaDisabled && !current && ((0, _keymap.is)('left', ev.keyCode) || (0, _keymap.is)('right', ev.keyCode))) {
          _this.paused.pause();

          _this.startListeningForPulses(ev.keyCode);
        }
      };

      _this.handleKeyUp = function (ev) {
        var _this$props2 = _this.props,
            mediaDisabled = _this$props2.mediaDisabled,
            no5WayJump = _this$props2.no5WayJump,
            rateChangeDisabled = _this$props2.rateChangeDisabled,
            playPauseButtonDisabled = _this$props2.playPauseButtonDisabled;
        if (mediaDisabled) return;

        if (!playPauseButtonDisabled) {
          if ((0, _keymap.is)('play', ev.keyCode)) {
            (0, _handle.forward)('onPlay', ev, _this.props);
          } else if ((0, _keymap.is)('pause', ev.keyCode)) {
            (0, _handle.forward)('onPause', ev, _this.props);
          }
        }

        if (!no5WayJump && ((0, _keymap.is)('left', ev.keyCode) || (0, _keymap.is)('right', ev.keyCode))) {
          _this.stopListeningForPulses();

          _this.paused.resume();
        }

        if (!rateChangeDisabled) {
          if ((0, _keymap.is)('rewind', ev.keyCode)) {
            (0, _handle.forward)('onRewind', ev, _this.props);
          } else if ((0, _keymap.is)('fastForward', ev.keyCode)) {
            (0, _handle.forward)('onFastForward', ev, _this.props);
          }
        }
      };

      _this.handleBlur = function () {
        _this.stopListeningForPulses();

        _this.paused.resume();
      };

      _this.handleWheel = function (ev) {
        if (_this.canShowMoreComponents() && _this.props.visible && ev.deltaY > 0) {
          _this.showMoreComponents();
        }
      };

      _this.startListeningForPulses = function (keyCode) {
        // Ignore new pulse calls if key code is same, otherwise start new series if we're pulsing
        if (_this.pulsing && keyCode !== _this.pulsingKeyCode) {
          _this.stopListeningForPulses();
        }

        if (!_this.pulsing) {
          _this.pulsingKeyCode = keyCode;
          _this.pulsing = true;
          _this.keyLoop = setTimeout(_this.handlePulse, _this.props.initialJumpDelay);
          (0, _handle.forward)('onJump', {
            keyCode: keyCode
          }, _this.props);
        }
      };

      _this.handlePulse = function () {
        (0, _handle.forward)('onJump', {
          keyCode: _this.pulsingKeyCode
        }, _this.props);
        _this.keyLoop = setTimeout(_this.handlePulse, _this.props.jumpDelay);
      };

      _this.handlePlayButtonClick = function (ev) {
        (0, _handle.forward)('onPlayButtonClick', ev, _this.props);

        if (_this.props.paused) {
          (0, _handle.forward)('onPlay', ev, _this.props);
        } else {
          (0, _handle.forward)('onPause', ev, _this.props);
        }
      };

      _this.getMediaControls = function (node) {
        if (!node) {
          _this.actionGuideHeight = 0;
          return;
        }

        _this.mediaControlsNode = _reactDom["default"].findDOMNode(node); // eslint-disable-line react/no-find-dom-node

        var guideElement = _this.mediaControlsNode.querySelector(".".concat(_MediaControlsModule["default"].actionGuide));

        _this.actionGuideHeight = guideElement ? guideElement.scrollHeight : 0;
      };

      _this.areMoreComponentsAvailable = function () {
        return _this.state.showMoreComponents;
      };

      _this.showMoreComponents = function () {
        _this.setState({
          showMoreComponents: true
        });
      };

      _this.hideMoreComponents = function () {
        _this.setState({
          showMoreComponents: false
        });
      };

      _this.handleClose = function (ev) {
        if (_this.props.visible) {
          (0, _handle.forward)('onClose', ev, _this.props);
        }
      };

      _this.handleFinish = function () {
        if (_this.state.showMoreComponents) {
          _this.paused.resume();

          if (!_spotlight["default"].getPointerMode()) {
            _spotlight["default"].move('down');
          }
        }
      };

      _this.handleCancel = function () {
        _this.paused.resume();
      };

      _this.mediaControlsNode = null;
      _this.moreComponentsNode = null;
      _this.actionGuideHeight = 0;
      _this.animation = null;
      _this.bottomComponentsHeight = 0;
      _this.keyLoop = null;
      _this.pulsingKeyCode = null;
      _this.pulsing = null;
      _this.paused = new _Pause["default"]('MediaPlayer');
      _this.state = {
        showMoreComponents: false,
        moreComponentsRendered: false
      };

      if (props.setApiProvider) {
        props.setApiProvider(_assertThisInitialized(_this));
      }

      return _this;
    }

    _createClass(MediaControlsDecoratorHOC, [{
      key: "componentDidMount",
      value: function componentDidMount() {
        (0, _dispatcher.on)('keydown', this.handleKeyDown, document);
        (0, _dispatcher.on)('keyup', this.handleKeyUp, document);
        (0, _dispatcher.on)('blur', this.handleBlur, window);
        (0, _dispatcher.on)('wheel', this.handleWheel, document);
      }
    }, {
      key: "componentDidUpdate",
      value: function componentDidUpdate(prevProps, prevState) {
        // Need to render `moreComponents` to show it. For performance, render `moreComponents` if it is actually shown.
        if (!prevState.showMoreComponents && this.state.showMoreComponents && !this.state.moreComponentsRendered) {
          this.moreComponentsRenderingJob.startRafAfter();
        } else if (prevState.showMoreComponents && !this.state.showMoreComponents) {
          this.moreComponentsRenderingJob.stop();
        }

        if (!prevState.moreComponentsRendered && this.state.moreComponentsRendered || this.state.moreComponentsRendered && prevProps.bottomComponents !== this.props.bottomComponents || !(0, _util2.compareChildren)(this.props.children, prevProps.children)) {
          this.calculateMoreComponentsHeight();
        }

        if (this.state.showMoreComponents && !prevState.moreComponentsRendered && this.state.moreComponentsRendered || this.state.moreComponentsRendered && prevState.showMoreComponents !== this.state.showMoreComponents) {
          forwardToggleMore({
            type: 'onToggleMore',
            showMoreComponents: this.state.showMoreComponents,
            liftDistance: this.bottomComponentsHeight - this.actionGuideHeight
          }, this.props);

          if (this.state.showMoreComponents) {
            this.moreComponentsNode = this.moreComponentsNode || this.mediaControlsNode.querySelector(".".concat(_MediaControlsModule["default"].moreComponents));
            this.paused.pause();
            this.animation = this.moreComponentsNode.animate([{
              transform: 'none',
              opacity: 0,
              offset: 0
            }, {
              transform: "translateY(".concat(-this.actionGuideHeight, "px)"),
              opacity: 1,
              offset: 1
            }], {
              duration: animationDuration,
              fill: 'forwards'
            });
            this.animation.onfinish = this.handleFinish;
            this.animation.oncancel = this.handleCancel;
          } else if (this.animation != null) {
            this.animation.cancel();
          }
        } // if media controls disabled, reset key loop


        if (!prevProps.mediaDisabled && this.props.mediaDisabled) {
          this.stopListeningForPulses();
          this.paused.resume();
        }
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        (0, _dispatcher.off)('keydown', this.handleKeyDown, document);
        (0, _dispatcher.off)('keyup', this.handleKeyUp, document);
        (0, _dispatcher.off)('blur', this.handleBlur, window);
        (0, _dispatcher.off)('wheel', this.handleWheel, document);
        this.stopListeningForPulses();
        this.moreComponentsRenderingJob.stop();

        if (this.animation) {
          this.animation.cancel();
        }
      }
    }, {
      key: "stopListeningForPulses",
      value: function stopListeningForPulses() {
        this.pulsing = false;

        if (this.keyLoop) {
          clearTimeout(this.keyLoop);
          this.keyLoop = null;
        }
      }
    }, {
      key: "toggleMoreComponents",
      value: function toggleMoreComponents() {
        this.setState(function (prevState) {
          return {
            showMoreComponents: !prevState.showMoreComponents
          };
        });
      }
    }, {
      key: "render",
      value: function render() {
        var props = Object.assign({}, this.props);
        delete props.initialJumpDelay;
        delete props.jumpDelay;
        delete props.moreActionDisabled;
        delete props.no5WayJump;
        delete props.onFastForward;
        delete props.onJump;
        delete props.onPause;
        delete props.onPlay;
        delete props.onRewind;
        delete props.onToggleMore;
        delete props.rateChangeDisabled;
        delete props.setApiProvider;
        return /*#__PURE__*/(0, _jsxRuntime.jsx)(Wrapped, _objectSpread(_objectSpread({
          ref: this.getMediaControls
        }, props), {}, {
          moreComponentsRendered: this.state.moreComponentsRendered,
          onClose: this.handleClose,
          onFlickFromActionGuide: this.handleFlickFromActionGuide,
          onKeyDownFromMediaButtons: this.handleKeyDownFromMediaButtons,
          onPlayButtonClick: this.handlePlayButtonClick,
          onTransitionEnd: this.handleTransitionEnd,
          showMoreComponents: this.state.showMoreComponents
        }));
      }
    }], [{
      key: "getDerivedStateFromProps",
      value: function getDerivedStateFromProps(props) {
        if (!props.visible) {
          return {
            showMoreComponents: false
          };
        }

        return null;
      }
    }]);

    return MediaControlsDecoratorHOC;
  }(_react.Component);

  MediaControlsDecoratorHOC.displayName = 'MediaControlsDecorator';
  MediaControlsDecoratorHOC.propTypes =
  /** @lends sandstone/MediaPlayer.MediaControlsDecorator.prototype */
  {
    /**
     * The label for the action guide.
     *
     * @type {String}
     * @public
     */
    actionGuideLabel: _propTypes["default"].string,

    /**
     * These components are placed below the children. Typically these will be media playlist items.
     *
     * @type {Node}
     * @public
     */
    bottomComponents: _propTypes["default"].node,

    /**
     * The number of milliseconds that the player will pause before firing the
     * first jump event on a right or left pulse.
     *
     * @type {Number}
     * @default 400
     * @public
     */
    initialJumpDelay: _propTypes["default"].number,

    /**
     * The number of milliseconds that the player will throttle before firing a
     * jump event on a right or left pulse.
     *
     * @type {Number}
     * @default 200
     * @public
     */
    jumpDelay: _propTypes["default"].number,

    /**
     * Disables the media buttons.
     *
     * @type {Boolean}
     * @public
     */
    mediaDisabled: _propTypes["default"].bool,

    /**
     * Disables showing more components.
     *
     * @type {Boolean}
     * @public
     */
    moreActionDisabled: _propTypes["default"].bool,

    /**
     * Setting this to `true` will disable left and right keys for seeking.
     *
     * @type {Boolean}
     * @public
     */
    no5WayJump: _propTypes["default"].bool,

    /**
     * Called when media fast forwards.
     *
     * @type {Function}
     * @public
     */
    onFastForward: _propTypes["default"].func,

    /**
     * Called when media jumps.
     *
     * @type {Function}
     * @public
     */
    onJump: _propTypes["default"].func,

    /**
     * Called when media gets paused.
     *
     * @type {Function}
     * @public
     */
    onPause: _propTypes["default"].func,

    /**
     * Called when media starts playing.
     *
     * @type {Function}
     * @public
     */
    onPlay: _propTypes["default"].func,

    /**
     * Called when media rewinds.
     *
     * @type {Function}
     * @public
     */
    onRewind: _propTypes["default"].func,

    /**
     * Called when the visibility of more components is changed
     *
    	 * Event payload includes:
     *
     * * `type` - Type of event, `'onToggleMore'`
     * * `showMoreComponents` - `true` when the components are visible`
     * * `liftDistance` - The distance, in pixels, the component animates
     *
     * @type {Function}
     * @public
     */
    onToggleMore: _propTypes["default"].func,

    /**
     * The video pause state.
     *
     * @type {Boolean}
     * @public
     */
    paused: _propTypes["default"].bool,

    /**
     * Disables state on the media "play"/"pause" button
     *
     * @type {Boolean}
     * @public
     */
    playPauseButtonDisabled: _propTypes["default"].bool,

    /**
     * Disables the media playback-rate control via rewind and fast forward keys
     *
     * @type {Boolean}
     * @public
     */
    rateChangeDisabled: _propTypes["default"].bool,

    /**
     * Registers the MediaControls component with an
     * [ApiDecorator]{@link core/internal/ApiDecorator.ApiDecorator}.
     *
     * @type {Function}
     * @private
     */
    setApiProvider: _propTypes["default"].func,

    /**
     * The visibility of the component. When `false`, the component will be hidden.
     *
     * @type {Boolean}
     * @public
     */
    visible: _propTypes["default"].bool
  };
  MediaControlsDecoratorHOC.defaultProps = {
    initialJumpDelay: 400,
    jumpDelay: 200
  };
  return (0, _Slottable["default"])({
    slots: ['bottomComponents']
  }, MediaControlsDecoratorHOC);
});
exports.MediaControlsDecorator = MediaControlsDecorator;

var handleCancel = function handleCancel(ev, _ref8) {
  var onClose = _ref8.onClose;

  if (onClose) {
    onClose(ev);
  }
};
/**
 * A set of components for controlling media playback and rendering additional components.
 *
 * This uses [Slottable]{@link ui/Slottable} to accept the custom tags, `<bottomComponents>`
 * to add components to the bottom of the media controls. Any additional children will be
 * rendered into the "more" controls area. Showing the additional components is handled by
 * `MediaControls` when the user navigates down from the media buttons.
 *
 * @class MediaControls
 * @memberof sandstone/MediaPlayer
 * @mixes ui/Cancelable.Cancelable
 * @ui
 * @public
 */


var MediaControls = (0, _ApiDecorator["default"])({
  api: ['areMoreComponentsAvailable', 'showMoreComponents', 'hideMoreComponents']
}, MediaControlsDecorator((0, _Cancelable["default"])({
  modal: true,
  onCancel: handleCancel
}, MediaControlsBase)));
exports.MediaControls = MediaControls;
MediaControls.defaultSlot = 'mediaControlsComponent';
var _default = MediaControls;
exports["default"] = _default;
